src/core/lib/surface/init.cc

```diff
@@ -225,10 +225,17 @@ void grpc_shutdown(void) {
 
 void grpc_shutdown_blocking(void) {
   GRPC_API_TRACE("grpc_shutdown_blocking(void)", 0, ());
-  grpc_core::MutexLock lock(&g_init_mu);
+  grpc_core::ReleasableMutexLock lock(&g_init_mu);
+  // it's ok to mark g_shutting_down at this time since the state only makes a
+  // difference in grpc_init when g_initializations is 0, at which time
+  // grpc_shutdown_internal_locked() must have already been called and hence
+  // g_shutting_down has been reset to false.
```

Comment:
This changes the behavior of the function and this is further away from the original grpc_shutdown function now. I think it might be better to add some comments in the header file and run all the tests internally as well.